packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,902 @@
+        }
+
+        // Create the calldata for the call
+        bytes memory txData = new bytes(4 + 32 + txDataSize);
+        assembly {
+            mstore(add(txData, 32), selector)
+            mstore(add(txData, 36), 0x20)
+            calldatacopy(add(txData, 68), offsetToGroups, txDataSize)
+        }
+
+        return txData;
+    }
+
+    function readTransfer(Context memory ctx)
+        internal
+        pure
+        returns (uint packedData, address to, address from)
+    {
+        // TransferTransaction.readTx(txsData, ctx.txIdx++ * ExchangeData.TX_DATA_AVAILABILITY_SIZE, transfer);
+
+        // Start by reading the first 23 bytes into packedData
+        uint txsDataPtr = ctx.txsDataPtr + 23;
+        // packedData: txType (1) | type (1) | fromAccountID (4) | toAccountID (4) | tokenID (2) | amount (3) | feeTokenID (2) | fee (2) | storageID (4)
+        assembly {
+            packedData := calldataload(txsDataPtr)
+            to := and(calldataload(add(txsDataPtr, 20)), 0xffffffffffffffffffffffffffffffffffffffff)
+            from := and(calldataload(add(txsDataPtr, 40)), 0xffffffffffffffffffffffffffffffffffffffff)
+        }
+        ctx.txsDataPtr += ExchangeData.TX_DATA_AVAILABILITY_SIZE;
+    }
+
+    function verifySignatureL2(
+        Context memory ctx,
+        address        owner,
+        uint           _accountID,
+        bytes32        txHash
+        )
+        internal
+        pure
+    {
+        /*
+        // Read the signature verification transaction
+        SignatureVerificationTransaction.SignatureVerification memory verification;
+        SignatureVerificationTransaction.readTx(txsData, ctx.txIdx++ * ExchangeData.TX_DATA_AVAILABILITY_SIZE, verification);
+
+        // Verify that the hash was signed on L2
+        require(
+            verification.owner == owner &&
+            verification.accountID == ctx.accountID &&
+            verification.data == uint(txHash) >> 3,
+            "INVALID_OFFCHAIN_L2_APPROVAL"
+        );
+        */
+
+        // Read the signature verification transaction
+        // Start by reading the first 25 bytes into packedDate
+        uint txsDataPtr = ctx.txsDataPtr + 25;
+        // packedData: txType (1) | owner (20) | accountID (4)
+        uint packedData;
+        uint data;
+        assembly {
+            packedData := calldataload(txsDataPtr)
+            data := calldataload(add(txsDataPtr, 32))
+        }
+
+        // Verify that the hash was signed on L2
+        require(
+            packedData & 0xffffffffffffffffffffffffffffffffffffffffffffffffff ==
+            (uint(ExchangeData.TransactionType.SIGNATURE_VERIFICATION) << 192) | ((uint(owner) & 0x00ffffffffffffffffffffffffffffffffffffffff) << 32) | _accountID &&
+            data == uint(txHash) >> 3,
+            "INVALID_OFFCHAIN_L2_APPROVAL"
+        );
+
+        ctx.txsDataPtr += ExchangeData.TX_DATA_AVAILABILITY_SIZE;
+    }
+
+    function encode(BridgeOperations calldata operations)
```

Comment:
do we need this?

Reply:
No, will remove once everything is final! I currently use this function to help encode the data that needs to be sent on-chain encodede in `bytes`. It's easier to do it like this if the function signature changes a lot in the tests, otherwise I had to update the encoding each time manually.